examples-gallery/plot_ball_rolling_on_spinning_disc.py

```diff
@@ -0,0 +1,620 @@
+iYY = iXX
+iZZ = iXX
+
+I = me.inertia(A1, iXX, iYY, iZZ)
+ball = me.RigidBody('ball', Dmc, A1, mb, (I, Dmc))
+observer = me.Particle('observer', m_Dmc, mo)
+bodies = [ball, observer]
+
+forces = [(Dmc, -mb*g*N.z), (m_Dmc, -mo*g*N.z),
+    (A1, t1*A1.x + t2*A1.y + t3*A1.z)]
+
+kd = sm.Matrix([ux - x.diff(t), uy - y.diff(t),
+    *[(rot - rot1).dot(uv) for uv in N ]])
+speed_constr = sm.Matrix([ux - r*u2, uy + r*u1])
+hol_constr = sm.Matrix([x - r*q2, y + r*q1])
+
+q_ind = [q1, q2, q3]
+q_dep = [x, y]
+u_ind = [u1, u2, u3]
+u_dep = [ux, uy]
+
+KM = me.KanesMethod(N,
+    q_ind=q_ind,
+    q_dependent=q_dep,
+    u_ind=u_ind,
+    u_dependent=u_dep,
+    kd_eqs=kd,
+    velocity_constraints=speed_constr,
+    configuration_constraints=hol_constr,
+    )
+
+(fr, frstar) = KM.kanes_equations(bodies, forces)
+eom = kd.col_join(fr + frstar)
+eom = me.msubs((eom.col_join(hol_constr)),
+    {sm.Derivative(T(t), t):
+    Tdot, sm.Derivative(T(t), (t,2)): Tdotdot},
+    )
+print(f'eom contains {sm.count_ops(eom):,} operations, ' +
+    f'{sm.count_ops(sm.cse(eom))} after cse')
+
+# %%
+# Set up the optimization problem and solve it.
+
+h = sm.symbols('h')
+state_symbols = (*q_ind, *q_dep, *u_ind, *u_dep)
+laenge = len(state_symbols)
+constant_symbols = (r, mb, mo, g, Omega, alpha, Tdot, Tdotdot)
+specified_symbols = (t1, t2, t3)
+methode = "backward euler"
+num_nodes = 250
+duration = (num_nodes - 1) * h
+
+# %%
+# Disc time is a constant, related to the acceleration of the disc.
+disc_time = 7.5
+
+# %%
+interval_value = h
+interval_value_fix = disc_time/num_nodes
+interval_fix = np.linspace(0, disc_time, num_nodes)
+
+# %%
+# Specify the known system parameters.
+par_map = {}
+par_map[mb]  = 5.0
+par_map[mo]  = 1.0
+par_map[r]   = 1.0
+par_map[Omega] = 10.0
+par_map[alpha] = 0.5
+par_map[g]   = 9.81
+par_map[Tdot] = interval_value_fix
+par_map[Tdotdot] = 0.0
+
+# %%
+# A weighted sum of energy and speed is to be minimized. weight is the
+# relative weight of the speed. (If weight = 0, only the energy is
+# minimized.)
+weight = 2.5e5
+
+# %%
+# Define the objective function and its gradient.
+def obj(free):
+    free1 = free[0: -1]
+    Tz1 = free1[laenge * num_nodes: (laenge + 1) * num_nodes]
+    Tz2 = free1[(laenge + 1) * num_nodes: (laenge + 2) * num_nodes]
+    Tz3 = free1[(laenge + 2) * num_nodes: (laenge + 3) * num_nodes]
+    return free[-1] * (np.sum(Tz1**2) + np.sum(Tz2**2) +
+        np.sum(Tz3**2)) + free[-1] * weight
+
+def obj_grad(free):
+    grad = np.zeros_like(free)
+    grad[laenge * num_nodes: (laenge + 1) * num_nodes] = (2.0 * free[-1]
+         * free[laenge * num_nodes: (laenge + 1) * num_nodes])
+    grad[(laenge + 1) * num_nodes: (laenge + 2) * num_nodes] = (2.0 *
+        free[-1] * free[(laenge + 1) * num_nodes: (laenge + 2) * num_nodes])
+    grad[(laenge + 2) * num_nodes: (laenge + 3) * num_nodes] = (2.0 * free[-1]
+       * free[(laenge + 2) * num_nodes: (laenge + 3) * num_nodes])
+    grad[-1] = weight
+    return grad
```

Comment:
These two functions are not easily understandable for an example.

Reply:
Can `create_objective_function` not be used?

Reply:
> Can `create_objective_function` not be used?

to the best of my knowledge, it cannot handle variable h. Am I wrong?

Reply:
Correct.